src/netlink.c

```diff
@@ -90,20 +90,27 @@ static int meth_send(lua_State *L) {
     size_t payload_size;
     const char *payload = luaL_checklstring(L, 2, &payload_size);
     int flags = luaL_optinteger(L, 3, 0);
-    struct nlmsgbuf nlb;
-    p_timeout tm = &nl->tm; 
+    p_timeout tm = &nl->tm;
     size_t sent = 0;
     int err;
+    struct nlmsgbuf *nlb = malloc(sizeof(struct nlmsgbuf));
```

Comment:
Have you measured the performance impact of this change? I think you should allocate this buffer only once in the Lua userdata and reuse it. Reusing shouldn't be a problem given that this function is not reentrant and Lua is single-threaded.

Reply:
We haven't had the time for a proper performance evaluation of this, no. I feel like it shouldn't be particularly significant, though, since there tends to be a lot of mallocing going on in Lua code anyway and this is in conjunction with a syscall + memcpying so we're doing something relatively expensive regardless.

Reply:
We are moving towards reducing the the memory usage in Lua; reusing tables and so on. In particular, we are want to reduce the overhead when receiving messages from the kernel, not add more overhead. If you don't want to bother measuring the change, you could at least perform a change that theoretically won't add overhead. If you still don't want to have this minimal amount of work, you could add an #ifdef that doesn't allocate the buffer in the heap by default and only allocate it there on the necessary platforms.

Reply:
Yeah, I do understand your point, and both the userdata and the `#ifdef` alternatives make sense to me.

For the `#ifdef` though we'd have to back and determine which exact platforms need this, which is unfortunately some manual work.

I'll have to get back to this later, once I have some spare time.

(Also I feel like it's worth pointing out that measuring this is not a minimal amount of work, since we don't have well-defined benchmarks for testing pretty much anything and microbenchmarks aren't always indicative of real-world performance, and we'd likely have to do this across several platforms which unfortunately isn't something we've automated to the point of convenience.)

Reply:
The minimal amount of work is moving these mallocs to the userdata structure, not measuring the performance. This change should be simple, just add another field to the t_netlink struct:
https://github.com/getCUJO/luasocket/blob/a0c55f75e1eaa4c825f40c8611e6111a1e70ccad/src/netlink.h#L31
You don't even need to add a malloc, since this object is already on the heap.
```
typedef struct {
    t_socket fd; 
    t_timeout tm;  
    t_pid srcpid; 
    t_type type; 
    struct nlmsgbuf nlmsgbuf; /* this should be very easy */
} t_netlink;
```

Reply:
Ah, then I misunderstood, apologies. I didn't realize there was already a struct like that for adding things to, I'm not very familiar with this code. Thanks for the hint, I can probably get that change done and tested soonish.

Reply:
Pushed an update with this type of change. I didn't squash it since I felt preserving James's authorship for the original change to be worth leaving it separate.